subsys/bluetooth/host/scan.c

```diff
@@ -279,99 +312,205 @@ static int start_le_scan_legacy(uint8_t scan_type, uint16_t interval, uint16_t w
+			.type = BT_LE_SCAN_TYPE_PASSIVE,
+			.options = BT_LE_SCAN_OPT_FILTER_DUPLICATE,
+			.interval = CONFIG_BT_BACKGROUND_SCAN_INTERVAL,
+			.window = CONFIG_BT_BACKGROUND_SCAN_WINDOW,
+			.timeout = 0,
+			.interval_coded = 0,
+			.window_coded = 0,
+		};
 
-	if (IS_ENABLED(CONFIG_BT_EXT_ADV) &&
-	    BT_DEV_FEAT_LE_EXT_ADV(bt_dev.le.features)) {
-		struct bt_hci_ext_scan_phy scan_phy_params;
-
-		scan_phy_params.type = BT_HCI_LE_SCAN_PASSIVE;
-		scan_phy_params.interval = sys_cpu_to_le16(interval);
-		scan_phy_params.window = sys_cpu_to_le16(window);
-
-		/* Scan on 1M + Coded if the controller supports it*/
 		if (BT_FEAT_LE_PHY_CODED(bt_dev.le.features)) {
-			return start_le_scan_ext(&scan_phy_params, &scan_phy_params, 0);
-		} else {
-			return start_le_scan_ext(&scan_phy_params, NULL, 0);
+			scan_param->options |= BT_LE_SCAN_OPT_CODED;
 		}
 
+		if (atomic_test_bit(scan_state.scan_flags, BT_LE_SCAN_USER_SYNC_SYNCING) ||
+		    atomic_test_bit(scan_state.scan_flags, BT_LE_SCAN_USER_SCAN_BEFORE_INITIATE)) {
+			scan_param->window = BT_GAP_SCAN_FAST_WINDOW;
+			scan_param->interval = BT_GAP_SCAN_FAST_INTERVAL;
+		}
 	}
-
-	return start_le_scan_legacy(BT_HCI_LE_SCAN_PASSIVE, interval, window);
 }
 
-int bt_le_scan_update(bool fast_scan)
+static int start_scan(struct bt_le_scan_param *scan_param)
 {
-	if (atomic_test_bit(bt_dev.flags, BT_DEV_EXPLICIT_SCAN)) {
-		/* The application has already explicitly started scanning.
-		 * We should keep the scanner running to avoid changing scan parameters.
-		 */
-		return 0;
+	if (IS_ENABLED(CONFIG_BT_EXT_ADV) && BT_DEV_FEAT_LE_EXT_ADV(bt_dev.le.features)) {
+		return start_le_scan_ext(scan_param);
 	}
 
-	if (atomic_test_bit(bt_dev.flags, BT_DEV_SCANNING)) {
-		int err;
+	return start_le_scan_legacy(scan_param);
+}
+
+static bool is_already_using_same_params(struct bt_le_scan_param *scan_param)
+{
+	return !memcmp(scan_param, &scan_state.used_scan_param, sizeof(*scan_param));
+}
+
+static int scan_update(void)
+{
+	int32_t err = 0;
+
+	struct bt_le_scan_param scan_param;
+
+	/* Prevent partial updates of the scanner state. */
+	k_mutex_lock(&scan_update_mutex, K_FOREVER);
+
+	select_scan_params(&scan_param);
+
+	enum scan_action {
+		SCAN_ACTION_NONE,
+		SCAN_ACTION_START,
+		SCAN_ACTION_STOP,
+		SCAN_ACTION_UPDATE,
+	};
 
-		err = bt_le_scan_set_enable(BT_HCI_LE_SCAN_DISABLE);
+	enum scan_action action = SCAN_ACTION_NONE;
+
+	bool is_scanning = atomic_test_bit(bt_dev.flags, BT_DEV_SCANNING);
+
+	/* Check if there is reason to have the scanner running */
+	if (atomic_get(scan_state.scan_flags) != 0) {
+		if (is_scanning) {
+			if (is_already_using_same_params(&scan_param)) {
+				/* Already scanning with the desired parameters */
+				action = SCAN_ACTION_NONE;
+			} else {
+				action = SCAN_ACTION_UPDATE;
+			}
+		} else {
+			action = SCAN_ACTION_START;
+		}
+	} else {
+		/* Scanner should not run */
+		if (is_scanning) {
+			action = SCAN_ACTION_STOP;
+		} else {
+			action = SCAN_ACTION_NONE;
+		}
```

Comment:
I think the compiler is smart enough to update the code accordingly. But I can remove the actions from the if statements.
I prefer to initialize the variable since that's less error-prone.